### ISSUE TITLE
Bug: [no-unnecessary-type-arguments] False positive with `Array.reduce`

### ISSUE THREAD

**Issue by TastyPi** [typescript-eslint/typescript-eslint#12196]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.9.3&fileType=.tsx&code=PTAEBUAsFNQBQDYEMCeBzATgewK4DsATUAraAZ1DywBdRoNsNQAjaAYyRzNgEtqByCmWo8ECUFjygAbgA4AdAFYA7PIBMAKA3UUAB1gAxLFlABeUAG8WSDAC5QwjDzxpQAXwDcWgu2QZYbJLCoDZ2oAAURiYAPqD4PgBmztAEAJQA2gC6XhogoACiAMoAMs60CchoFAAGUdUhFPh47ORkNigANCw4tP4AtljSzq58oBxc5La5YOCFAIyyAAwALLL2-Lr%2B0vygPBS6WGRkPMwIKKD88dBJzQT88hqh8v4EOGzQADxRAHzhGqARTbQaRdNg4BipMzfCJWZg2exghgAfnkcKYSKRoCB0lRNncqQ6-0s1jCACIAB6k9yE1IeIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWiIDmsALaIWqDJETRoHaJHBgAviF1A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDArugDTg2R4BDTAHN0LSpHQkATAAZp0yDxpR4sACbsMrSTPmLlfNZowAlKbFTsCxHQCN2RdRkyReAX0ORkRaFfXorADavLSY6IiYsAB0eGS8ALog7kA&tokens=false

### Repro Code

```TypeScript
type Foo = { bar: string };

declare const arr: (Foo | undefined)[];

// ESLint flags `Foo` as unnecessary, but removing it causes:
// TS18048: 'prev' is possibly 'undefined'.
arr.reduce<Foo>(
  (prev, curr) => ({ bar: curr?.bar ?? prev.bar }),
  { bar: "x" },
);
```

### ESLint Config

```javascript
import eslint from "@eslint/js";
import * as tseslint from "typescript-eslint";

export default tseslint.config(
  tseslint.configs.base,
  {
    files: ["**/*.ts"],
    languageOptions: {
      parserOptions: {
        projectService: true,
      },
    },
    rules: {
      "@typescript-eslint/no-unnecessary-type-arguments": "error",
    },
  },
);
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true,
    "target": "ES2022",
    "module": "ES2022",
    "moduleResolution": "bundler"
  },
  "include": ["repro.ts"]
}
```

### Expected Result

eslint should not tell me to remove the type parameter, as it is inferred as `Foo | undefined` not `Foo`.

### Actual Result

`eslint` tells me to remove the type parameter, then `tsc` complains that `prev` can be `undefined`, when it can't.

**Comment by bradzacher** [comment]
This error is not showing in the playground.
Could you please create an isolated reproduction.

**Comment by TastyPi** [comment]
@bradzacher I have updated the playground link with the failure

**Comment by bradzacher** [comment]
Duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/12158

**Comment by TastyPi** [comment]
@bradzacher I don't think this is the same issue though, as it was not a problem in 8.57.1. Unless it's the same issue that has somehow got worse.
